### ISSUE TITLE
Can I add more images or data to the model already done custom training?

### ISSUE THREAD

**Issue by B7Kompirine** [ultralytics/ultralytics#1631]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I'm made an uav detection model with 500 images. I just tried 128 epochs which took about 3 hours of training. I like the way it is working but I want to train more to get close to perfect because it has some difficult time to recognize with different background color/environment/surroundings.

The question is, if I want to simply add more images on the same model to train more, how can I do that? Is there a way that I can load custom trained model? Like can I take the best.py output and paste it to command then start training with more image/label file that I prepared?

Since it takes so long with 500 images I can't imagine how long it will take with 3,000 images. I want to continue training model that has already done trained.

To be more simple, I want to continue to train best.pt with more labels and images

Thank you for your helps...

Please note: This is my first time doing it and I am Aircraft Engineer. Thats why please explain in details and threat me like a beginner while teaching me how because maybe I can't understand what you trying to explain.

Please don't hesitate to ask more information.  :)

### Additional

This is the line that I am using: `yolo task=detect mode=train epochs=128 data=data_custom.yaml model=yolov8m.pt imgsz=640 batch=11 patience=64`

And this is the folder with best.py: `C:\Users\musti\OneDrive\Desktop\TheCoding\YOLOV8\runs\detect\train2\weights\best.pt`

**Comment by glenn-jocher** [comment]
@B7Kompirine you can start training from any pretrained model:

```
yolo train model=path/to/any/model.pt
```

**Comment by B7Kompirine** [comment]
> 

Will this model be my output right? I am talking about best.pt because when I started training I used yolov8m.pt which is pre model from ultralytics.

**Comment by glenn-jocher** [comment]
@B7Kompirine older models are not overwritten by new trainings. New run directories are created for logging new training results including training checkpoints.

**Comment by B7Kompirine** [comment]
@glenn-jocher then do I have to start training everything from beginning? Can't I use already trained model and add more features to it? Like as I said I want to improve the quality of it but I don't want to train it from 0

**Comment by B7Kompirine** [comment]
@glenn-jocher thank you I did it. I copied the best.pt and paste it to main folder with different name and I change the model name as you said and it did worked.

`yolo task=detect mode=train epochs=128 data=data_custom.yaml model=**yolov8m.pt** imgsz=640 batch=11 patience=64`

`yolo task=detect mode=train epochs=128 data=data_custom.yaml model=**yolov8m_custom_train1.pt** imgsz=640 batch=11 patience=64`

**Comment by glenn-jocher** [comment]
Great!

**Comment by SheeppLi** [comment]
I wonder what is changed when I train on custom dataset, the whole network weight or the last part? when I run on custom dataset and get the best.pt, and transfer it to another custom dataset to get another best.pt, will the network forget what it learns in coco or the first custom dataset? 
In fact when I trained on another dataset which has the same label with the first one, I find that the code shows transfer number less than the total number, it makes me confused.

**Comment by glenn-jocher** [comment]
@SheeppLi all parameters with gradients are updated during training.

**Comment by stphtan94117** [comment]
@glenn-jocher 
So if I use yolov5 instead of yolov8, is the reason the same?
Can I use best.pt to train the second model?

**Comment by glenn-jocher** [comment]
Yes you can use best.pt to train subsequent models. You can do this either YOLOv5 or YOLOv8.

**Comment by shahbazahmed108** [comment]
> @glenn-jocher thank you I did it. I copied the best.pt and paste it to main folder with different name and I change the model name as you said and it did worked.
> 
> `yolo task=detect mode=train epochs=128 data=data_custom.yaml model=**yolov8m.pt** imgsz=640 batch=11 patience=64`
> 
> `yolo task=detect mode=train epochs=128 data=data_custom.yaml model=**yolov8m_custom_train1.pt** imgsz=640 batch=11 patience=64`

i hope someone can read this, i am trying to acheive the same thing. but the issue is that when i label the new dataset, i have same classes as my pre-trained model and new classes. the index of the classes don't match as in my pre-trained model label dataset and new label dataset. does this cause issue, if yes, how to solve this

**Comment by glenn-jocher** [comment]
@shahbazahmed108 when adding new classes to an existing model, ensure that the class indices in your new dataset match those from the pre-trained model's dataset. If they don't align, you'll need to update the class indices in your new dataset to match the pre-trained model's class indices for the existing classes, and then append the new classes at the end. This will prevent any confusion during training and ensure the model learns correctly.

**Comment by glenn-jocher** [comment]
@shahbazahmed108 yes, you can use the existing `best.pt` to continue training on your new dataset with classes [B,C,D,E]. Just ensure that the class indices for B and C remain consistent with the original training, and append D and E afterwards. This process is known as transfer learning and can help improve performance on the new classes without starting from scratch.

**Comment by shahbazahmed108** [comment]
@glenn-jocher I trained yolov8 model on images with different classes and the .yaml for that data set was like this

train: C:\Users\name\Desktop\leica\train
val: C:\Users\name\Desktop\leica\val

nc: 13

names: ["ARG","WM","CW","FW-TM","CL","MW","KOCHIA","PL","PL-AS","PW","RTF","BROME","THISTLE"]

this training gave me the best.pt and that was running great on all 13 classes

then i have more images with a few classes as in the above training group and also some new classes. what I wanted to do is grab the already trained model as in the above dataset and train that further with a new dataset that shares some old classes and has some new ones. the number of classes for this dataset was 10 as PW, MW, PL, UB, CL, WM, WHEAT, HENBIT, NIGHTSHADE, BT. all these images are sitting in a new folder different from already trained model. that folder name is field_images - corrected-final

as you see some of the classes are new and some are old as in my first dataset. so to train the existing model with the already trained weight file name as best.pt, I created a new .yaml file with common classes with both datasets matching their indices and new classes were added at the end. so my new .yaml file for training looked like this

train: C:\Users\Shahbaz\Desktop\field_images - corrected-final\train
val: C:\Users\Shahbaz\Desktop\field_images - corrected-final\val

nc: 18

names: ["ARG","WM","CW","FW-TM","CL","MW","KOCHIA","PL","PL-AS","PW","RTF","BROME","THISTLE","UB","WHEAT","HENBIT","NIGHTSHADE","BT"]

remember the idea was to grab the best.pt file from the first trained model and train it further on new images which had some old classes and some new ones. for the second .yaml file, I arranged classes such that the old classes matched their place as old training and new classes were appended at the end. the problem is that when I ran the training on a new dataset hoping that the new best.pt will detect all old and new classes, it did not work and it is only detecting new classes from 18. the command that I used looked like this

yolo task=detect mode=train epochs=100 data=data_custom.yaml model="C:\Users\Shahbaz\Desktop\leica\best.pt"  imgsz=640

The training process went fine but the detection is not happening as I hoped. 
Please help me how to fix this

**Comment by glenn-jocher** [comment]
@shahbazahmed108 it sounds like you've done the right steps by aligning the class indices of the common classes and appending the new ones. However, if the model is only detecting the new classes, it's possible that the class indices or names may not be perfectly aligned between the two datasets. Double-check the indices in your new `.yaml` file to ensure they match exactly with those in the original dataset for the common classes. Also, ensure that the `task` in your command is set to `train` and not `detect` when you're training with the new dataset. If the issue persists, consider training with a combined dataset of old and new images to reinforce the learning of the old classes.

**Comment by aramirr** [comment]
Hi @glenn-jocher, I trained a model  with 500 images and classes A, B, C and D.  I'm not getting the best results for Recall in classes C and D, so I guess I need to train the model with more images of C and D, let's say I take 300 more images.

When re-training the best.pt model that I got from the previous process, do I need to train it using the original 500 images + 300 new images? Or should I use only the new 300 images?

In therms of epochs, what is the best option? Use the same amount of epochs as for the first process? or it doesn't matter?

Thanks!!

**Comment by glenn-jocher** [comment]
@aramirr hello! Great questions! 👍

When improving class-specific performance, it's often beneficial to retrain with both the original and new images to maintain the model's performance on all classes. So, yes, using the 500 original images plus the additional 300 focusing on classes C and D would be ideal.

As for epochs, continuing to train for a similar number of epochs as your initial training can be a good starting point. Adjust based on the performance improvements you observe. More isn't always better, so monitor for signs of overfitting.

Here's how you could set it up:
```bash
yolo task=detect mode=train epochs=your_preferred_epochs data=your_dataset.yaml model=best.pt imgsz=640
```
Adjust `epochs` and `imgsz` as needed!

Hope that helps! Let me know how it goes. 😊

**Comment by aramirr** [comment]
Hi @glenn-jocher,

Thank you so much for your answer and quick reply! :) I will re-tran the model during this weekend.

This leads me to a second question.

Would it be the same to train a model with X classes and Y images with 600 epochs, than to train a model-B with the same X classes and Y images but with 300 epochs, and then re-train the same model-B with the same X classes and Y images for 300 more epochs (600 epochs in total)??

Thanks and regards!

**Comment by glenn-jocher** [comment]
Hi there!

Great question! Training a model for 600 epochs straight versus training it for 300 epochs, and then retraining for another 300 epochs can have different effects. Continuous training for 600 epochs allows the model to learn and adjust its weights without interruptions. Retraining in sessions (like the 300 + 300 split) could potentially provide "breaks" where the model sort of consolidates what it has learned before further tuning in the next session.

However, you might not see a big difference if the data doesn't change between sessions. It is generally okay to split the training if it's more manageable for your workflow or computational resources.

Remember to monitor the performance metrics and adjust accordingly!

Happy training! 😊
